lib/apollo-federation/tracing/tracer.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+module ApolloFederation
+  module Tracing
+    module Tracer
+      # store string constants to avoid creating new strings for each call to .trace
+      EXECUTE_QUERY = 'execute_query'
+      EXECUTE_QUERY_LAZY = 'execute_query_lazy'
+      EXECUTE_FIELD = 'execute_field'
+      EXECUTE_FIELD_LAZY = 'execute_field_lazy'
+
+      def self.trace(key, data, &block)
+        case key
+        when EXECUTE_QUERY
+          execute_query(data, &block)
+        when EXECUTE_QUERY_LAZY
+          execute_query_lazy(data, &block)
+        when EXECUTE_FIELD
+          execute_field(data, &block)
+        when EXECUTE_FIELD_LAZY
+          execute_field_lazy(data, &block)
+        else
+          yield
+        end
```

Comment:
I'm confused about the implementation of the tracer. I couldn't find any documentation of these four events (`execute_query`, `execute_query_lazy`, `execute_field`, `execute_field_lazy`) beyond just their names here:
https://graphql-ruby.org/api-doc/1.9.11/GraphQL/Tracing
So I don't understand what the difference is between the lazy and non-lazy ones and whether the tracer is tracking each one correctly. Can you shed more light on how you figured this out?

The "Step 1" "Step 2"... comments are also confusing - is there some documentation that guarantees the events always occur in that order?

Finally - why is the `execute_query` event used to record only a `start_time` timestamp and not an `end_time` timestamp after the `block.call` line? And similarly why is `execute_query_lazy` only recording the `end_time`? It seems like we're assuming that both of these events always occur in sequence. Is there documentation for that?

Reply:
@noaelad TBH, I probably wouldn't have figured this out on my own. I used this in-progress PR as reference: https://github.com/salsify/apollo-tracing-ruby/blob/feature/new-apollo-api/lib/apollo_tracing/tracer.rb

I believe the steps are guaranteed to execute in the order described in my comments. The naming is definitely confusing—might have to poke rmosolgo for an explaination.

Reply:
My concern is around relying on undocumented behavior in `graphql-ruby`, especially where that undocumented behavior doesn't make a lot of immediate sense to me.

Reply:
Have you tried setting both `start_time` and `end_time` in both `execute_query` and `execute_query_lazy`?

Reply:
I pulled your branch and tried playing around, when both methods (`execute_query` and `execute_query_lazy`) have the same implementation the only spec failures result from the fact that timestamps are increased twice and not just once (which makes sense in today's world). I'd prefer an implementation like this that is more robust and relies less on the fact that `graphql-ruby` happens to call both `execute_query` and `execute_query_lazy` in a specific sequence. WDYT?

By the way the specs would still be a brittle to changes in `graphql-ruby` but I don't have a good suggestion for mitigating that.

```
      def self.execute_query(data, &block)
        query = data.fetch(:query)
        return block.call unless query.context && query.context[:tracing_enabled]

        trace = query.context.namespace(ApolloFederation::Tracing::KEY)
        trace.merge!(
          start_time: Time.now.utc,
          start_time_nanos: Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond),
          node_map: NodeMap.new,
        )
        result = block.call
        trace.merge!(
          end_time: Time.now.utc,
          end_time_nanos: Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond),
        )

        result
      end

      def self.execute_query_lazy(data, &block)
        execute_query(data, &block)
      end
```

Reply:
So this was a fun rabbit hole, but I think I understand tracing mechanics well now. I added a fat comment that explains the ordering of trace events. The key was figuring out that there are two execution phases: 1) non-lazy, 2) lazy. I could not figure out a way for  `execute_query_lazy` to not fire, so I think it's still a good place to record ending times.

[Here's a graphql-ruby test](https://github.com/rmosolgo/graphql-ruby/blob/6b2138d5904eee89f2420b37cf99679a15f2bf84/spec/graphql/tracing/platform_tracing_spec.rb) that shows execution order. It always has `eql` as the last step in the event list.

We could record an end time at the end of `execute_query` but it will always get thrown away.

Reply:
Thanks for digging into this